File: code/Object-C/SSNSDK/SSNSDK/NSSocketManager.m
//
//  NSSocketManager.m
//  Test
//
//  Created by 阿杰 on 2018/4/2.
//  Copyright © 2018年 阿杰. All rights reserved.
//
#import "NSMessageHead.h"
#import "NSSocketManager.h"
static NSSocketManager *manager;
static NSInteger version;
@implementation NSSocketManager{
    BOOL isConnectSuccess,canSendMesssage,isInputStreamOpenSuccess,isOutputStreamOpenSuccess,logOpen;
}
+(NSSocketManager *)manager{
    if (!manager) {
        manager = [[NSSocketManager alloc]init];
        version = 0x01;
    }
    return manager;
}

-(void)openLog:(BOOL)isopen{
    logOpen = isopen;
}

-(NSMutableArray *)chatMsgs{
    
    if(!_chatMsgs) {
        
        _chatMsgs =[NSMutableArray array];
        
    }
    
    return _chatMsgs;
}

-(void)stream:(NSStream *)aStream handleEvent:(NSStreamEvent)eventCode{
    if (logOpen) {
        NSLog(@"%@",[NSThread currentThread]);
    }
   
    
    //NSStreamEventOpenCompleted = 1UL << 0,//输入输出流打开完成//NSStreamEventHasBytesAvailable = 1UL << 1,//有字节可读//NSStreamEventHasSpaceAvailable = 1UL << 2,//可以发放字节//NSStreamEventErrorOccurred = 1UL << 3,//连接出现错误//NSStreamEventEndEncountered = 1UL << 4//连接结束
    
    switch(eventCode) {
            
        case NSStreamEventOpenCompleted:
            if (aStream == self.inputStream) {
                isInputStreamOpenSuccess = YES;
            }else if (aStream == self.outputStream){
                isOutputStreamOpenSuccess = YES;
            }
            if (isInputStreamOpenSuccess && isOutputStreamOpenSuccess) {
                if (logOpen) {
                    NSLog(@"输入输出流打开完成");
                }
                
                isConnectSuccess = YES;
                if (self.delegate && [self.delegate respondsToSelector:@selector(onSocketConnectWithResult:)]) {
                    [self.delegate onSocketConnectWithResult:YES];
                }
            }
            break;
        case NSStreamEventHasBytesAvailable:
            {if (logOpen) {
                NSLog(@"有字节可读");
            }
            NSData * data =[self readData];
            if ([NSMessageHead checkHeadByData:data]) {
                if (self.delegate && [self.delegate respondsToSelector:@selector(onSocketGetMessage:)]) {
                    [self.delegate onSocketGetMessage:[data subdataWithRange:NSMakeRange(8, data.length - 8)]];
                }
            }
            break;}
            
        case NSStreamEventHasSpaceAvailable:
            if (logOpen) {
                NSLog(@"可以发送字节");
            }
            
            canSendMesssage = YES;
            if (self.delegate && [self.delegate respondsToSelector:@selector(onSocketCanSendMessage)]) {
                [self.delegate onSocketCanSendMessage];
            }
            break;
            
        case NSStreamEventErrorOccurred:
            if (logOpen) {
                NSLog(@"连接出现错误");
            }
            
            if (aStream == self.inputStream) {
                isInputStreamOpenSuccess = NO;
            }else if (aStream == self.outputStream){
                isOutputStreamOpenSuccess = NO;
            }
            isConnectSuccess = NO;
            if (self.delegate && [self.delegate respondsToSelector:@selector(onSocketConnectWithResult:)]) {
                [self.delegate onSocketConnectWithResult:NO];
            }
            break;
            
        case NSStreamEventEndEncountered:
            if (logOpen) {
                NSLog(@"连接结束");
            }
            
            isConnectSuccess = NO;
            canSendMesssage = NO;
            isInputStreamOpenSuccess = NO;
            isOutputStreamOpenSuccess = NO;
            //关闭输入输出流
            
            [self.inputStream close];
            
            [self.outputStream close];
            
            //从主运行循环移除
            
            [self.inputStream removeFromRunLoop:[NSRunLoop mainRunLoop] forMode:NSDefaultRunLoopMode];
            
            [self.outputStream removeFromRunLoop:[NSRunLoop mainRunLoop] forMode:NSDefaultRunLoopMode];
            if (self.delegate && [self.delegate respondsToSelector:@selector(onSocketCloseConnect)]) {
                [self.delegate onSocketCloseConnect];
            }
            break;
            
        default:
            
            break;
            
    }
    
}


-(void)connectToHost:(NSString *)host andPort:(int)port{
    
    //定义C语言输入输出流
    
    CFReadStreamRef readStream;
    
    CFWriteStreamRef writeStream;
    
    CFStreamCreatePairWithSocketToHost(NULL, (__bridge CFStringRef)host, port, &readStream, &writeStream);
    //把C语言的输入输出流转化成OC对象
    
    self.inputStream = (__bridge NSInputStream *)(readStream);
    
    self.outputStream = (__bridge NSOutputStream *)(writeStream);
    
    //设置代理
    
    self.inputStream.delegate=self;
    
    self.outputStream.delegate=self;
    
    //把输入输入流添加到主运行循环
    
    //不添加主运行循环 代理有可能不工作
    
    [self.inputStream scheduleInRunLoop:[NSRunLoop mainRunLoop] forMode:NSDefaultRunLoopMode];
    
    [self.outputStream scheduleInRunLoop:[NSRunLoop mainRunLoop] forMode:NSDefaultRunLoopMode];
    
    //打开输入输出流
    
    [self.inputStream open];
    
    [self.outputStream open];
    
}

#pragma mark 读了服务器返回的数据

-(NSData *)readData{
    
    //建立一个缓冲区 可以放1024个字节
    NSMutableData * msgs = [NSMutableData data];
    while ([self.inputStream hasBytesAvailable]) {
        uint8_t buf[1024];
        
        //返回实际装的字节数
        
        NSInteger len = [self.inputStream read:buf maxLength:sizeof(buf)];
        
        //把字节数组转化成字符串
        
        NSData *data =[NSData dataWithBytes:buf length:len];
        
        //从服务器接收到的数据
        [msgs appendData:data];
    }
    return msgs;
}
-(void)sendMessage:(NSData *)msg{
    if (msg && isConnectSuccess && canSendMesssage) {
        NSMessageHead * head = [[NSMessageHead alloc]init];
        head.length = msg.length + 8;
        head.type = 0x01;
        head.version = version;
        NSMutableData * data = [NSMutableData dataWithData:[head getHeadData]];
        [data appendData:msg];
        [self.outputStream write:data.bytes maxLength:data.length];
    }else{
        if (logOpen) {
            NSLog(@"Error:Connect has not connected");
        }
    }
    
}
@end

File: code/Swift/SSNSDK/SSNSDK/NSSocketManager.swift
//
//  NSSocketManager.swift
//  SSNSDK
//
//  Created by 阿杰 on 2018/5/2.
//  Copyright © 2018年 阿杰. All rights reserved.
//

import Foundation
protocol NSSocketDelegate:NSObject{
    func onSocketConnectWithResult(isSuccess:Bool) -> Void
    func onSocketGetMessage(messages:Data) -> Void
    func onSocketCanSendMessage() -> Void
    func onSocketCloseConnect() -> Void
}

class NSSocketManager: NSObject,StreamDelegate {
    static let manager: NSSocketManager = {
        return NSSocketManager()
    }()
    static let version: Int32 = 0x01
    var isConnectSuccess:Bool = false,canSendMesssage:Bool = false,isInputStreamOpenSuccess:Bool = false,isOutputStreamOpenSuccess:Bool = false,logOpen:Bool = true;

    var inputStream: InputStream! {
        get {
            return self.inputStream
        }
        set {
            self.inputStream = newValue
        }
    }
    var outputStream: OutputStream! {
        get {
            return self.outputStream
        }
        set {
            self.outputStream = newValue
        }
    }
    var delegate: NSSocketDelegate!
    
    func connectToHost(host:String,port:UInt32) -> Void {
        //定义C语言输入输出流
    
        var readStream:Unmanaged<CFReadStream>?
    
        var writeStream:Unmanaged<CFWriteStream>?
        CFStreamCreatePairWithSocketToHost(nil, host as CFString, port, &readStream, &writeStream)
        //把C语言的输入输出流转化成OC对象
        if readStream != nil {
            self.inputStream = readStream!.takeRetainedValue()
        }
    
        if writeStream != nil {
            self.outputStream = writeStream!.takeRetainedValue()
        }
    
        //设置代理
    
        self.inputStream.delegate=self;
    
        self.outputStream.delegate=self;
    
        //把输入输入流添加到主运行循环
    
        //不添加主运行循环 代理有可能不工作
        self.inputStream.schedule(in: RunLoop.main, forMode: RunLoopMode.defaultRunLoopMode)
        self.outputStream.schedule(in: RunLoop.main, forMode: RunLoopMode.defaultRunLoopMode)
    
        //打开输入输出流
    
        self.inputStream.open()
    
        self.outputStream.open()
    }
    func openLog(isopen:Bool) -> Void {
        logOpen = isopen
    }
    func stream(_ aStream: Stream, handle eventCode: Stream.Event) {
        if (logOpen) {
            print(Thread.current)
        }
        
        switch eventCode {
        case Stream.Event.openCompleted:
            if (aStream == self.inputStream) {
                isInputStreamOpenSuccess = true;
            }else if (aStream == self.outputStream){
                isOutputStreamOpenSuccess = true;
            }
            if (isInputStreamOpenSuccess && isOutputStreamOpenSuccess) {
                if (logOpen) {
                    print("输入输出流打开完成");
                }
                
                isConnectSuccess = true;
                if (self.delegate != nil && self.delegate.responds(to: Selector.init(("onSocketConnectWithResult:")))) {
                    self.delegate.onSocketConnectWithResult(isSuccess: true)
                }
            }
            break
        case Stream.Event.hasBytesAvailable:
            if (logOpen) {
                print("有字节可读");
            }
            let data = self.readData()
            if (NSMessageHead.checkHeadByData(data: data)) {
                if (self.delegate != nil && self.delegate.responds(to: Selector.init(("onSocketGetMessage:")))) {
                self.delegate.onSocketGetMessage(messages:data.subdata(in: 8..<data.count - 8));
                }
            }
            break
        case Stream.Event.hasSpaceAvailable:
            if (logOpen) {
                print("可以发送字节");
            }
            
            canSendMesssage = true;
            if (self.delegate != nil && self.delegate.responds(to: Selector.init(("onSocketCanSendMessage"))))  {
                self.delegate.onSocketCanSendMessage()
            }
            break
        case Stream.Event.errorOccurred:
            if (logOpen) {
                print("连接出现错误");
            }
            if aStream == self.inputStream {
                isInputStreamOpenSuccess = false
            }else if aStream == self.outputStream {
                isOutputStreamOpenSuccess = false
            }
            isConnectSuccess = false
            if (self.delegate != nil && self.delegate.responds(to: Selector.init(("onSocketConnectWithResult:")))) {
                self.delegate.onSocketConnectWithResult(isSuccess: false)
            }
            break
        case Stream.Event.endEncountered:
            if logOpen {
                print("连接结束")
            }
            isConnectSuccess = false
            canSendMesssage = false
            isOutputStreamOpenSuccess = false
            isInputStreamOpenSuccess = false
            self.inputStream.close()
            self.outputStream.close()
            self.inputStream.remove(from: RunLoop.main, forMode: RunLoopMode.defaultRunLoopMode)
            self.outputStream.remove(from: RunLoop.main, forMode: RunLoopMode.defaultRunLoopMode)
            if (self.delegate != nil && self.delegate.responds(to: Selector.init(("onSocketCloseConnect")))){
                self.delegate.onSocketCloseConnect()
            }
            break
        default:
            break
        }
    }


    func readData() -> Data {
        var data = Data.init()
        while self.inputStream.hasBytesAvailable {
            let bytes:UnsafeMutablePointer<UInt8> = UnsafeMutablePointer<UInt8>.allocate(capacity: 1024)
            let len = self.inputStream.read(bytes, maxLength:1024)
            let tmpData:Data = Data.init(bytes: bytes, count: len)
            data.append(tmpData)
        }
        return data
    }

    func sendMessage(msg:Data) -> Void {
        if (isConnectSuccess && canSendMesssage) {
            let head:NSMessageHead = NSMessageHead.init();
            head.length = Int32(msg.count) + 8;
            head.type = 0x01;
            head.version = NSSocketManager.version;
            var data:Data = Data.init();
            data.append(head.getHeadData())
            data.append(msg);
            let bytes:UnsafeMutablePointer<UInt8> = UnsafeMutablePointer<UInt8>.allocate(capacity: data.count)
            data.copyBytes(to: bytes, count: data.count)
            self.outputStream.write(bytes, maxLength: data.count)
        }else{
            if (logOpen) {
                print("Error:Connect has not connected")
            }
        }
    }
}
